3scale/3scale-operator
# Add support for PodDisruptionBudget

## WHAT

This PR will add support for PodDisruptionBudget to the 3scale operator. See [JIRA](https://issues.redhat.com/browse/THREESCALE-4258).

## HOW

A new feature flag is introduced to the `APIManager` CRD to allow users to enable this feature.  An example can be found [here](https://github.com/wei-lee/3scale-operator/blob/ae4397ca2585fabdc57ce83cef118bfb51357e35/deploy/crds/apps_v1alpha1_apimanager_cr_pdb.yaml).

When enabled, the following PDBs are created:

* apicast-production
* apicast-staging
* backend-cron
* backend-listener
* backend-worker
* system-app
* system-sidekiq
* zync
* zync-que

All PDBs have the `MaxUnavailable` field value set to `1`. This will ensure when the node is being drained, only 1 pod per component is allowed to be evicted.  We do not to update/delete these PDBs even if there is only 1 instance of these components are deployed as it won't violate the rules in PDBs. Also all the components can still be scaled up and down as normal. It will be count against the PDB rules, but PDB won't stop replica controllers from scaling down the pods.

## Verification

To verify:

1. Follow the instructions [here](https://github.com/3scale/3scale-operator/blob/master/doc/development.md#deploy-3scale-operator-manually) to build the operator and deploy it an OpensShift 4 cluster. You can use this image `quay.io/wei_lee/3scale-operator:test`.
2. Create a new APIManager CR using [this example](https://github.com/wei-lee/3scale-operator/blob/ae4397ca2585fabdc57ce83cef118bfb51357e35/deploy/crds/apps_v1alpha1_apimanager_cr_pdb.yaml).
3. Wait for deployments start for all the components.
4. List all the PDBs in the namespace and you should see 9 of them. You can also check the content of them to make sure they have the right spec.
5. Try scale down some of the components and you should still be able to do that.
6. Delete the APIManager CR that is created and make sure all the PDBs are deleted as well.
